pallets/settlement/src/lib.rs

```diff
@@ -1225,7 +1344,16 @@ impl<T: Config> Module<T> {
 
         // Lock tokens for all legs that are not of type [`Leg::OffChain`]
         let filtered_legs = Self::filtered_legs(instruction_id, &portfolios_set);
-        for (leg_id, leg) in filtered_legs.leg_subset() {
+        // If the fee was estimated in advance, the input values must be at least equal to the actual values
+        if let Some(input_cost) = input_cost {
+            Self::ensure_valid_input_cost(
+                filtered_legs.sender_asset_count(),
+                filtered_legs.receiver_asset_count(),
+                filtered_legs.unfiltered_asset_count().off_chain(),
+                &input_cost,
+            )?;
```

Comment:
Could just pass a reference to `filtered_legs` to `ensure_valid_input_cost`, since all calls to that function take values from the `filtered_legs` object.